vllm_spyre/v1/core/scheduler.py

```diff
@@ -523,28 +594,126 @@ def schedule(self) -> "SchedulerOutput":
 
         return outputs
 
-    def can_schedule(self, request):
-        # For currently prefilling requests: all the required blocks were
-        # already reserved and all the conditions were already checked at the
-        # time of scheduling the first chunk. We also always prioritize prefill
-        # over decode, and this allows us to skip verification of later chunks.
-        #
-        # Soon, we will add some constraints here:
-        # 1. If it is not the last chunk: schedule the prefill if we have enough
-        #    blocks for a single chunked prefill, and if previous step was a
-        #    decode step (we interleave prefills with decodes)
-        # 2. If it is the last chunk: do a prefill if all the other scheduling
-        #    conditions are met
+    def can_schedule_prefill(self, request: Request) -> bool:
+        # running and waiting queues are both empty, we can start a new batch
+        # which can always be scheduled
+        if len(self.running) + len(self.waiting) == 0:
+            return True
+
+        if not self._has_scheduling_priority(request):
+            return False
+
+        return self._satisfies_constraints(request)
+
+    def _satisfies_constraints(self, request: Request) -> bool:
+        is_first_chunk = request.num_computed_tokens == 0
+        is_last_chunk = (request.num_prompt_tokens -
+                         request.num_computed_tokens) <= self.chunk_size
```

Comment:
general comment here is that it looks like we have separate sets of constraints for
1. accepting a new prefill (is_first_chunk)
2. starting decodes for the last prefill (is_last_chunk)

Can we separate these into different methods to make it clear which is which?

```
can_schedule = True
if is_first_chunk:
    can_schedule = _satisfies_prefill_constraints(req)
if is_last_chunk and can_schedule:
    can_schedule = _satisfies_decode_constraints(req)

return can_schedule
```